src/python/WMCore/ReqMgr/Service/Request.py

```diff
@@ -423,37 +426,19 @@ def _handleNoStatusUpdate(self, workload, request_args, dn):
             cherrypy.log('Updated workqueue statistics of "{}", with:  {}'.format(workload.name(), reqArgs))
             return report
 
-        reqArgsNothandled = []
-        for reqArg in reqArgs:
-            if reqArg == 'RequestPriority':
-                validate_request_priority(reqArgs)
-                # must update three places: GQ elements, workload_cache and workload spec
-                self.gq_service.updatePriority(workload.name(), reqArgs['RequestPriority'])
-                workload.setPriority(reqArgs['RequestPriority'])
-                cherrypy.log('Updated priority of "{}" to: {}'.format(workload.name(), reqArgs['RequestPriority']))
-            elif reqArg == "SiteWhitelist":
-                workload.setSiteWhitelist(reqArgs["SiteWhitelist"])
-                cherrypy.log('Updated SiteWhitelist of "{}", with:  {}'.format(workload.name(), reqArgs['SiteWhitelist']))
-            elif reqArg == "SiteBlacklist":
-                workload.setSiteBlacklist(reqArgs["SiteBlacklist"])
-                cherrypy.log('Updated SiteBlacklist of "{}", with:  {}'.format(workload.name(), reqArgs['SiteBlacklist']))
-            else:
-                reqArgsNothandled.append(reqArg)
-                cherrypy.log("Unhandled argument for no-status update: %s" % reqArg)
-
-        reqStatus = self.reqmgr_db_service.getRequestByNames(workload.name())[workload.name()]['RequestStatus']
-        cherrypy.log(f"CurrentRequest status: {reqStatus}")
-        if reqArgsNothandled:
+        try:
+            # Commit all Global WorkQueue changes per workflow in a single go:
+            self.gq_service.updateElementsByWorkflow(workload, reqArgs, status=['Available', 'Negotiating', 'Acquired'])
```

Comment:
The `WMWorkloadUnhandledException` is telling you if there have been any unhandled  arguments. It is thrown when we call  `workload.updateWorkloadArgs` from within `updateElementsByWorkflow` - here: https://github.com/dmwm/WMCore/pull/12099/files#diff-d758570ce7f6baddaeefb0bdb4e2015b06e5b7772bf929248b686adc64286233R316 

And I did it so because of your request during our discussion of the issue to have the `workflow` and the `workqueue elements` being updated in a single push. 
It was not like that just until last night. And I've asked you at least twice ... trying to make all the needed pointers in my comments here: https://github.com/dmwm/WMCore/pull/12099#discussion_r1876282940 and here:  https://github.com/dmwm/WMCore/pull/12099#issuecomment-2539343491 before I proceed with unifying those two steps with this commit:  https://github.com/dmwm/WMCore/pull/12099/commits/93d0c3b0513074795b500711b0414eef599191a1 (hence why I still keep all the commits not squashed)

Reply:
And I do not see how would we recognize if there have been any workqueue elements to be updated until we  query the the workqueue and see if there are any. It have been sequential step and a mandatory one in this method even before I make it all in one go. We  were still always going through it, as long as we were called.  

Usually the parameter `reqArgsDiff` was the one that would have told us if there was anything to be done at this stage within this method, but I remember in the past you asked me to avoid calling the validation twice at the head of this method here. But regardless... , in all the cases of a call to update a request with no `RequestStatus` update we still always return the `reqArgsDiff` from here: https://github.com/dmwm/WMCore/blob/44d2a8a77a42f885bb99f3c4128b823b4e1929ea/src/python/WMCore/ReqMgr/Utils/Validation.py#L111 instead of the originally passed `request_args` and then any actions should have been halted at the very top of this method here: https://github.com/dmwm/WMCore/blob/44d2a8a77a42f885bb99f3c4128b823b4e1929ea/src/python/WMCore/ReqMgr/Service/Request.py#L417-L419 

So  if there was nothing to be updated in the workflow itself - ergo in its `WQE` elements of the given status, we should never reach that point in the code. But if there have been any change with the workflow arguments instead, we will have to proceed with updating the `WQE` elements as well. 

On top of that we update only the ones in the selected statuses, rather than all of them. I think we are safe in what we are doing here.

Reply:
Given that these changes are now being applied on top of your previous incomplete changes, it is a bit harder to see how things were implemented.

I don't really understand why we call `_handleAssignmentStateTransition()` from this `_handleNoStatusUpdate()` method, as it says that there is not **supposed** to be any status transition. Can you please explain that?

In addition, I believe that this call `updateElementsByWorkflow()` could become not super cheap (few secs?) if workqueue is loaded with documents. So, IMO we should only call it if really necessary. To put it in a different way, a request getting assigned (hence in `assignment-approved`) does not have workqueue elements.

Reply:
> I don't really understand why we call _handleAssignmentStateTransition() from this _handleNoStatusUpdate() method, as it says that there is not supposed to be any status transition. Can you please explain that? 

Because, if you have called the `_handleNoStatusUpdate`  for a workflow that is in `assignment_approved` then you will get  with the request all those full set of parameters that you get during the `assignment_approved` state transition. Which will blow up, because we decided to still support a limited set of parameters for no status update - only `SiteLists` related and `RequestPriority` (which was yet again a wrong decision). This is the root cause of the BUG you reported in the first implementation an did not accept the code back then (even though I have provided this same solution on the very next day).  So this was the solution - the way out of the situation was - in order to avoid rewriting all which is already there in `_handleAssignmentStateTRansition` and assimilating it in  `workload.updateWorkloadArgs` and since the `_handleAssignmentStateTRansition` method has nothing to do with the state transition action itself.... it was safe enough to call it directly here. If you want, I will take this method and will rewrite it under the context of [`workload.updateWorkloadArgs`](https://github.com/dmwm/WMCore/pull/12099/files#diff-b045f03c5ee4e99d47b54e3e08582c4d1a23764cb0b7ba509734628b19c22d86R86). I do not mind. Actually I think this is the proper way to do it. And I've expressed that in the past.

Reply:
And what is the relationship of a workflow in `assignment_approved` with `updateElementsByWorkflow()`? None, the workflow has not been assigned yet. Calling workqueue is a waste of resources.

If we had an issue (maybe that was with ACDC), making a call to WorkQueue won't resolve that issue. This implementation is wrong. If you want to keep it, keep it.

Reply:
Alan, make up your mind..... 

you asked me to reduce the calls  to worload  and push it all in one go. I twisted my mind to implement it. Before I did it I asked for your feedback twice.... now you tell me this is wrong.

Reply:
If we want to have it all correct we must stop putting exrta background knowledge in the algorithms we follow ... and tie ourselves in a `figure 8` knot of conditions.

Reply:
In order to avoid confusion and blame .... Please put the correct set of conditions here in simple English under which the work queue update elements must be called.  And I will implement it.

Reply:
As discussed over Zoom:
* `_handleAssignmentStateTransition` is not performing any status transition;
* and the exception `WMWorkloadUnhandledException` is not meant to be raised by `updateElementsByWorkflow()`, but by its internal call `workload.updateWorkloadArgs(updateParams)`, if needed.

Thank you for explaining this Todor. I have no more concerns on this code.